pydm/data_store/__init__.py

```diff
@@ -0,0 +1,182 @@
+            The key for the engineering unit information at the data dictionary
+        precision_key : str
+            The key for the precision information at the data dictionary
+        upper_limit_key : str
+            The key for the upper limit information at the data dictionary
+        lower_limit_key : str
+            The key for the lower limit information at the data dictionary
+
+        Returns
+        -------
+        introspection : dict
+
+        """
+        lookup_table = [
+            (connection_key, DataKeys.CONNECTION),
+            (value_key, DataKeys.VALUE),
+            (severity_key, DataKeys.SEVERITY),
+            (write_access_key, DataKeys.WRITE_ACCESS),
+            (enum_strings_key, DataKeys.ENUM_STRINGS),
+            (unit_key, DataKeys.UNIT),
+            (precision_key, DataKeys.PRECISION),
+            (upper_limit_key, DataKeys.UPPER_LIMIT),
+            (lower_limit_key, DataKeys.LOWER_LIMIT)
+        ]
+        introspection = dict()
+
+        for val, key in lookup_table:
+            if val:
+                introspection[key] = val
+
+        return introspection
+
+
+DEFAULT_INTROSPECTION = {k: v for k, v in DataKeys.__dict__.items()
+                         if isinstance(v, str) and not k.startswith('_')}
+
+
+class DataStore(object):
+    """
+    Singleton class responsible for holding the data and instrospection tables
+    for the channels.
+    """
+    __instance = None
+
+    def __init__(self):
+        if self.__initialized:
+            return
+        self.__initialized = True
+        self._data = {}
+        self._introspection = {}
+
+    def __new__(cls, *args, **kwargs):
+        if cls.__instance is None:
+            cls.__instance = object.__new__(DataStore)
+            cls.__instance.__initialized = False
+        return cls.__instance
+
+    def introspect(self, address):
+        """
+        Query the introspection mapping about the information for a given
+        address.
+
+        Parameters
+        ----------
+        address : str
+            The address identifier.
+
+        Returns
+        -------
+        introspection : dict or None
+            If no information is found this method returns None
+        """
+        return self._introspection.get(address, None)
+
+    def fetch(self, address, with_introspection=False):
+        """
+        Fetch the data associated with an address.
+
+        Parameters
+        ----------
+        address : str
+            The address identifier
+        with_introspection : bool
+            If True, returns also the introspection dictionary
+
+        Returns
+        -------
+        data : dict or None
+            If no information is found this method returns None
+        introspection : dict
+            Returned only if `with_introspection` is `True`
+        """
+        data = self._data.get(address, {})
+        intro = self._introspection.get(address, {})
+        if with_introspection:
```

Comment:
Return type instability based on keyword arguments can be confusing/frustrating for users. Consider splitting this to `fetch` and `fetch_with_introspection`.

Reply:
Addressed at latest commit.